source/MRMesh/MRPointsSave.cpp

```diff
@@ -124,74 +131,127 @@ tl::expected<void, std::string> toCtm( const PointCloud& points, std::ostream& o
     {
         colors4f.resize( colors->size() );
         for ( int i = 0; i < colors4f.size(); ++i )
-            colors4f[i] = Vector4f( ( *colors )[VertId{i}] );
+            colors4f[i] = Vector4f( ( *colors )[VertId{ i }] );
 
-        ctmAddAttribMap( context, (const CTMfloat*) colors4f.data(), "Color" );
+        ctmAddAttribMap( context, ( const CTMfloat* )colors4f.data(), "Color" );
     }
 
     if ( ctmGetError( context ) != CTM_NONE )
         return tl::make_unexpected( "Error encoding in CTM-format colors" );
 
-    ctmSaveCustom( context, []( const void* buf, CTMuint size, void* data )
+    struct SaveData
     {
-        std::ostream& s = *reinterpret_cast<std::ostream*>( data );
-        s.write( (const char*) buf, size );
-        return s.good() ? size : 0;
-    }, &out );
+        std::function<bool( float )> callbackFn{};
+        std::ostream* stream;
+        size_t sum{ 0 };
+        size_t blockSize{ 0 };
+        size_t maxSize{ 0 };
+    } saveData;
+    saveData.callbackFn = [callback, &saveData] ( float progress )
+    {
+        // calculate full progress in partical-linear scale (we don't know compressed size and it less than real size)
+        progress = ( saveData.sum + progress * saveData.blockSize ) / saveData.maxSize;
+        float newProgress = 0.f;
+        for ( int i = 0; i < 5; ++i )
+        {
+            if ( progress < 0.2f )
+            {
+                newProgress += progress / 0.2f * 0.7f * ( 1 - newProgress );
+                break;
+            }
+            else
+            {
+                progress = ( progress - 0.2f ) / 0.8f;
+                newProgress += ( 1 - newProgress ) * 0.7f;
```

Comment:
Please introduce named constants (fewer - better) instead of 0.2, 0.7, 0.8

Reply:
removed magic numbers